src/EditorFeatures/TestUtilities/TestExportJoinableTaskContext.cs

```diff
@@ -1,24 +1,18 @@
 ﻿using System;
-using System.Collections.Generic;
 using System.ComponentModel.Composition;
-using System.Linq;
-using System.Text;
-using System.Threading;
-using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.VisualStudio.Threading;
 
 namespace Microsoft.CodeAnalysis.Editor.UnitTests
 {
-    // In 15.3 the editor took a dependency on JoinableTaskContext.
-    // JTC appears in the VS MEF composition but does not itself
-    // contain an ExportAttribute. The Editor's own unit tests
-    // export it from a field and we need to do the same in order
-    // to be able to compose in any part of the Editor.
-    internal class TestExportJoinableTaskContext : ForegroundThreadAffinitizedObject
+    // Starting with 15.3 the editor took a dependency on JoinableTaskContext
+    // in Text.Logic and Intellisense layers as an editor host provided service.
+    [PartCreationPolicy(CreationPolicy.NonShared)] // JTC is "main thread" affinitized so should not be shared
+    internal class TestExportJoinableTaskContext
     {
+        [ThreadStatic]
+        private static JoinableTaskContext s_jtcThreadStatic;
```

Comment:
It looks like you're solving the problem twice over: you're both using NonShared *and* you're using ThreadStatic here. 
It seems ThreadStatic by itself should fix this, shouldn't it?

Reply:
Nope, just ThreadStatic is not enough, tons of tests fail because JTC.IsOnMainThread check fails. 
@AArnott, what's the creation semantics of an exported property? I believe if the class is a shared part it's instantiated only once and its property is accessed only once too so we have a singleton JTC per catalog.
Roslyn tests [reuse ExportProvider](https://github.com/dotnet/roslyn/blob/6847f1e5a909395aae9456e8f366cbf4deb86b69/src/EditorFeatures/TestUtilities/TestExportProvider.cs#L27) containing this JTC export so when a test runs on a different thread JTC.IsOnMainThread return false. That's why I thought just having CreationPolicy.NonShared would solve it. But for some reason that alone deadlocks ([dump]("\\olegt-vmhost\tmp\xunit.runner.worker.dmp")) and I don't fully understand why yet. I suspect we end up creating more than one JTC targeting the same thread as main they get stored in various editor components in that static ExportProvider. That's why ThreadStatic.
Maybe a better way would be not have JTC export part of that static ExportProvider at all.

Reply:
> what's the creation semantics of an exported property?

An exported property's getter is called for each and every importer. Basically the same thing as an exported type when it's tagged as a NonShared part.

My first guess about the deadlock is that different importers are getting unique instances of JTC even within the same test and MEF graph, which means they can't interact to resolve deadlocks. If you're all on the same (main) thread, this probably wouldn't actually be an issue. But if you actually have multiple threads going for an individual test, and you care about the simulated 'main' thread, then you need to ensure that you have one JTC instance shared for the whole test and you need a way for a non-main thread in the test to schedule work for the 'main' thread. This would require that you set up a SynchronizationContext on the 'main' thread that can schedule work, and then see that that thread actually pumps those messages. It's not hard, but it must be done within your test project.

> Roslyn tests reuse ExportProvider

That's unfortunate. ExportProvider itself is actually really cheap. It's composing the graph that's expensive. I wonder if they've tried just constructing the catalog and graph once, and sharing the resulting ExportProviderFactory for all tests, but each individual test gets its own ExportProvider? That sounds like it would be the best combination of perf and test isolation.
But if you *must* share the ExportProvider, then we're basically saying that all tests need to share a JTC and thus share a 'main' thread. So there should be a static thread that's running a message pump for all tests to occasionally switch to during the test.

Reply:
The way I solved this in VsVim is to do the following:

- Define a new ExportProvider for JoinableTaskFactory https://github.com/jaredpar/VsVim/blob/master/Src/VimEditorHost/EditorHostFactory.JoinableTaskContextExportProvider.cs
- Add that to the ExportProviderList of the composition https://github.com/jaredpar/VsVim/blob/master/Src/VimEditorHost/EditorHostFactory.cs#L80

Given that all tests run on the main STA thread htis means the provider, and the resulting JoinableTaskContext, will be created on the main thread.

Reply:
Oh, I haven't realized the export was on a field, not a property. After turning it into a property I could remove part creation policy altogether, just leaving ThreadStatic as you suggested.

@jasonmalinowski, does the last iteration look reasonable enough to proceed with this PR?

Reply:
thanks, @jaredpar. I'm trying to follow that approach. Roslyn needs VS MEF ExportProvider though, so I'm trying to figure out how to create VS MEF Export provider for JTC and how to combine 2 VS MEF export providers.

Reply:
I guess it's not possible with VS MEF and switching to MEF v1 ExportProvider (like in VsVim) is not possible due to portable requirements. So the approach in this PR is the only realistic one I can think of.

@jasonmalinowski please review if I addressed the changes you requested in this PR.

Reply:
> I wonder if they've tried just constructing the catalog and graph once, and sharing the resulting ExportProviderFactory for all tests, but each individual test gets its own ExportProvider? That sounds like it would be the best combination of perf and test isolation.

Roslyn has switched back and forth between sharing the actual ExportProvider and just sharing the catalog around 5 times now. It alternated between "oh goodness, the perf!" and "oh goodness, we're sharing test state!" that it made me wonder if everybody was suffering from memory loss. :smile: I want to say last time the switch was done was prior to us moving to VS MEF (and might predate VS MEF itself....?), and at the time creating a new composition while reusing the catalog was cheap, but once you have 50,000 tests doing that and it's more expensive than the five lines of actual code being tested, you start to get second thoughts. :smile:

@olegtk I'm OK with this going in. I suspect it's still going to bite us eventually but I think that trap has already been set...

Awhile back I tried giving a shot at decoupling all the MEF shared state to also be strict about tests that are free-threaded (and can run in parallel) and tests that need a concept of a UI thread and it was a terrible rat hole. We had some tests that'd casually spin up a WPF text view on a random xunit test thread and weren't careful about that, or other sorts of fun. One Day(tm) I want to get it cleaned up, but you aren't making it any worse.

Reply:
@jaredpar I'd like to understand why you had to derive types for ExportProvider and/or JoinableTaskContext. You can 'inject' a JoinableTaskContext into a VS MEF composition using an exporting field/property and avoid any type deriving.